File: palettier-core/src/main/java/be/asmolabs/palettier/core/backup/BackupModel.java
package be.asmolabs.palettier.core.backup;

import java.time.Instant;
import java.util.List;

/**
 * Forme de la sauvegarde, telle qu'elle est ecrite dans le fichier JSON.
 *
 * <p>Volontairement independante des entites : une sauvegarde doit rester lisible et
 * reimportable meme apres un remaniement du modele. Les tubes y sont designes par leur
 * marque et leur nom, jamais par un identifiant de base -- un numero de ligne ne veut
 * rien dire dans une autre installation.</p>
 */
public final class BackupModel {

    /** Version du format, pour qu'une relecture ulterieure sache a quoi elle a affaire. */
    public static final int FORMAT_VERSION = 1;

    private BackupModel() {
    }

    public record Backup(int formatVersion,
                         Instant exportedAt,
                         String application,
                         List<Paint> paints,
                         List<Palette> palettes,
                         List<Project> projects,
                         List<Recipe> recipes) {
    }

    /**
     * Un tube tel que le peintre l'a chez lui : ses corrections comprises.
     *
     * @param tintHex teinte coupee de blanc, ou {@code null} si elle n'est pas connue
     * @param owned   vrai si le peintre declare le posseder
     */
    public record Paint(String brand, String name, String code, List<String> pigments,
                        String hex, String tintHex, String opacity, String dryingClass,
                        double tintingStrength, boolean owned, boolean colorDerived, String notes) {
    }

    /** Une palette, ses tubes designes par marque et nom. */
    public record Palette(String name, String purpose, String notes, List<PaintRef> paints) {
    }

    /** Renvoi vers un tube, par sa cle naturelle. */
    public record PaintRef(String brand, String name) {
    }

    public record Project(String name, String subject, String approach, String notes,
                          String paletteName, Instant createdAt,
                          List<PaintRef> paints, List<Zone> zones, List<Photo> photos) {
    }

    public record Zone(String name, String material, String note, List<Layer> layers) {
    }

    /**
     * Une couche : ce qui est vise, et ce qui a deja ete pose.
     *
     * <p>Les cinq derniers champs decrivent la pose. Ils sont absents des archives
     * ecrites avant qu'elle soit suivie, et valent alors {@code null} : la couche est
     * simplement rendue a peindre, ce qu'elle etait. Aucune raison de changer la version
     * du format pour cela -- une archive ancienne se relit sans perte.</p>
     *
     * @param kind       LADDER pour une marche du degrade, ACCENT pour une variation locale
     * @param appliedAt  date de pose, ou {@code null} si la couche reste a peindre
     */
    public record Layer(String role, String targetHex, String technique, String note, String kind,
                        Instant appliedAt, Double appliedTemperature, Double appliedHumidity,
                        String appliedVentilation, String appliedDryingClass) {
    }

    /**
     * Une photo, rangee a cote du JSON plutot qu'encodee dedans.
     *
     * @param file chemin du fichier dans l'archive, par exemple "images/projet-1-photo-2.jpg"
     */
    public record Photo(String file, String role, String caption, Instant addedAt) {
    }

    public record Recipe(String name, String subject, String notes, List<Step> steps) {
    }

    public record Step(String technique, String paintMix, String medium,
                       double mediumRatio, String thickness, String note) {
    }
}

File: palettier-kmp/core-data/src/commonMain/kotlin/be/asmolabs/palettier/data/backup/BackupModel.kt
package be.asmolabs.palettier.data.backup

import kotlinx.serialization.Serializable

/**
 * La forme de la sauvegarde ecrite par l'application Java, reproduite a l'identique.
 *
 * <p>C'est le pont entre les deux applications, et la raison pour laquelle aucun
 * convertisseur H2 vers SQLite n'a eu a etre ecrit : l'ancienne exporte, la nouvelle
 * importe. Le format avait ete concu independant des entites -- "une sauvegarde doit
 * rester lisible et reimportable meme apres un remaniement du modele" -- et c'est
 * exactement ce remaniement-la qui arrive.</p>
 *
 * <p>Les dates sont declarees en texte et analysees ensuite. Le Java les ecrit en
 * ISO-8601 ; les lire comme telles evite de dependre d'un serialiseur de date dont le
 * format pourrait differer d'une version a l'autre.</p>
 */
@Serializable
data class BackupDocument(
    val formatVersion: Int = 1,
    val exportedAt: String? = null,
    val application: String? = null,
    val paints: List<BackupPaint> = emptyList(),
    val palettes: List<BackupPalette> = emptyList(),
    val projects: List<BackupProject> = emptyList(),
    val recipes: List<BackupRecipe> = emptyList(),
)

@Serializable
data class BackupPaint(
    val brand: String,
    val name: String,
    val code: String = "",
    val pigments: List<String> = emptyList(),
    val hex: String,
    val tintHex: String? = null,
    val opacity: String,
    val dryingClass: String,
    val tintingStrength: Double = 0.5,
    val owned: Boolean = false,
    val colorDerived: Boolean = false,
    val notes: String = "",
)

/** Renvoi vers un tube, par sa cle naturelle. */
@Serializable
data class BackupPaintRef(val brand: String, val name: String)

@Serializable
data class BackupPalette(
    val name: String,
    val purpose: String = "",
    val notes: String = "",
    val paints: List<BackupPaintRef> = emptyList(),
)

@Serializable
data class BackupProject(
    val name: String,
    val subject: String = "",
    val approach: String = "",
    val notes: String = "",
    val paletteName: String? = null,
    val createdAt: String? = null,
    val paints: List<BackupPaintRef> = emptyList(),
    val zones: List<BackupZone> = emptyList(),
    val photos: List<BackupPhoto> = emptyList(),
)

@Serializable
data class BackupZone(
    val name: String,
    val material: String = "",
    val note: String = "",
    val layers: List<BackupLayer> = emptyList(),
)

/**
 * Une couche : ce qui est vise, et ce qui a deja ete pose.
 *
 * <p>Les cinq derniers champs sont absents des archives ecrites avant que la pose soit
 * suivie. La couche revient alors simplement a peindre, ce qu'elle etait.</p>
 */
@Serializable
data class BackupLayer(
    val role: String,
    val targetHex: String,
    val technique: String = "",
    val note: String = "",
    val kind: String = "LADDER",
    val appliedAt: String? = null,
    val appliedTemperature: Double? = null,
    val appliedHumidity: Double? = null,
    val appliedVentilation: String? = null,
    val appliedDryingClass: String? = null,
)

/** Une photo, rangee a cote du JSON plutot qu'encodee dedans. */
@Serializable
data class BackupPhoto(
    val file: String,
    val role: String,
    val caption: String = "",
    val addedAt: String? = null,
)

@Serializable
data class BackupRecipe(
    val name: String,
    val subject: String = "",
    val notes: String = "",
    val steps: List<BackupStep> = emptyList(),
)

@Serializable
data class BackupStep(
    val technique: String,
    val paintMix: String = "",
    val medium: String,
    val mediumRatio: Double = 0.0,
    val thickness: String,
    val note: String = "",
)
